### DISCUSSION TITLE
Collect built-in Accelerator performance counter

### DISCUSSION THREAD

**Discussion by LeiZhou-97** [sustainable-computing-io/kepler#681] [Ideas]
Currently, I'm working on collecting AMX (the next generation of AVX512) performance counters. On Intel SPR CPU, it has the following hw counter.

[EXE.AMX_BUSY](https://perfmon-events.intel.com/index.html?pltfrm=spxeon.html&evnt=EXE.AMX_BUSY): Counts the cycles where the AMX (Advance Matrix Extension) unit is busy performing an operation.

Based on this counter, we can get how long  has AMX been used.

Previously, I tried to use existing bpf framework in kepler to collect amx_busy like cpu_cycles. But I found this doesn't work for AMX. 

Cycles/cache_miss are synchronous hw counters. The counter values between task_switch belong the current running process.

For AMX, after CPU offloads data to AMX, the CPU will be occupied by other processes when AMX is running. So the counter value is counted on other processes.

Now my idea is to use Linux perf directly to track the amx-related counter values generated by this pid. I know it must have some overhead, but in bpf program to trace PMU event also depends on linux perf subsystem. So I personally think this is acceptable, and a config can be added to the kepler later to let the end user decide whether to turn on this feature, there is a trade-off.

Is this solution acceptable to the community? And everyone is welcome to suggest your better ideas. Thanks!

**Comment by marceloamaral** [comment]
Hey @LeiZhou-97, thank you for sharing this interesting use case!

The reason we prefer BPF over Linux perf is because Linux perf has more overhead.

The great thing is that any PMU event that Linux perf  can collect, BPF can collect as well.

In Kepler, we currently have code in two places for collecting performance events: the BPF code [here](https://github.com/sustainable-computing-io/kepler/blob/main/bpfassets/perf_event/perf_event.c#L56) and the golang code [here](https://github.com/sustainable-computing-io/kepler/blob/main/pkg/bpfassets/attacher/bcc_attacher.go#L58).

Each performance event is associated with a hexadecimal code. We use the "golang.org/x/sys/unix" library, which provides a list of widely available perf events, as seen [here](https://cs.opensource.google/go/x/sys/+/master:unix/linux/types.go;l=1131?q=PERF_COUNT_HW_CPU_CYCLES&sq=&ss=go%2Fx%2Fsys). However, since your case is more specific, you'll need to utilize the Linux perf to get the code, and then create a const in our code with this value.

Additionally, we might need to update the power mode to incorporate this new counter.

Could you please create a pull request (PR)? That would be really helpful!

**Reply by LeiZhou-97** [reply]
Sorry for my late response. 

My current implementation is already based on what you say.

```go
// RAW Performance Counter
// amx_busy: <umask=0x2><event=0xb7>
AMXBusyLabel: {unix.PERF_TYPE_RAW, 0x2b7, isSupportAMX()},
```

I added an new raw performance counter into the existing `perf_event.c`.

```
kepler_container_amx_busy_total{container_name="amx_pod"} 0
kepler_container_amx_busy_total{container_name="other_pods"} 1643803
kepler_container_amx_busy_total{container_name="system_processes"} 2031584854573
```

According to the result， you can see all amx events are counted on other pods and amx_pod is zero. (Since AMX is running, the CPU is occupied by other tasks) So that's the reason why I raise this discussion.

And I compare the between current kepler bpf implementation and linux perf.

**kepler:** SYS_perf_event_open (pid == -1 and cpu >= 0)

**linux perf per process:** SYS_perf_event_open (pid > 0 and cpu == -1)

They both rely on `perf_event_open`. 

Kernel does not have a dedicated func to trace AMX ops, therefore, I proposed to trace AMX all the time. (and add one config or policy for end user to decide if enable this feature.)

**Reply by LeiZhou-97** [reply]
Let me add some more comments.

AMX doesn't depends on the specific kfunc, user can even use AMX in userspace.

Sample code: https://github.com/intel/AMX-TMUL-Code-Samples/blob/main/src/test-amxtile.c

**Reply by marceloamaral** [reply]
When an application begins using the AMX accelerator, it requests access permission through a kernel call using the following syntax: syscall(SYS_arch_prctl, ARCH_REQ_XCOMP_PERM, XFEATURE_XTILEDATA). The kernel function is possibly related to this call [here](https://elixir.bootlin.com/linux/latest/source/arch/x86/um/syscalls_64.c#L16).
Then we can determine when the process starts utilizing the accelerator using this kernel function.

When the application finishes using the AMX, it releases the accelerator using the function `_tile_release()`. Although not entirely certain, it is likely that this function calls the kernel function [tile_release](https://elixir.bootlin.com/linux/latest/source/arch/x86/include/asm/special_insns.h#L277).

**Reply by LeiZhou-97** [reply]
Yep, you are right. Using `arch_prctl` can be seen if the current process has permission to use AMX.

But it cannot determine if AMX is being used. As I said before, the actual use of AMX does not depend on the kernel.

We can use `arch_prctl` as the config for checking if AMX is being used by the current process, and use Linux Perf for HW counter collection.

**Reply by marceloamaral** [reply]
Can multiple processes utilize AMX simultaneously?

If one process obtains access permission through `arch_prctl`, can another process also acquire the permission, or does the previous process need to release it first?

**Reply by LeiZhou-97** [reply]
Can multiple processes utilize AMX simultaneously?
[Yes, AMX is an in-die accelerator.]

If one process obtains access permission through arch_prctl, can another process also acquire the permission
[Yes]

 or does the previous process need to release it first?
[Not need, and I also bpftrace the func `tile_release()`, it does not be called during AMX running. The function annotation said it will be invoked from the cpuidle driver and prevent from entering low-power idle]

Then I thought about it again, and I take back what I said before that `arch_prctl` can be used as AMX trace enabling config. I think it's better to use a displaced configuration for end user. Because some of them may not need to trace AMX events when they use it, as we all know, tracing AMX events all the time has overhead.

**Comment by jiere** [comment]
Thanks @marceloamaral for quick reply, actually this topic is related to the latest discussion [here](https://github.com/sustainable-computing-io/kepler/issues/645), we plan to introduce a in-die(built-in) accelerator framework in current Kepler power source, before that, we need to first figure out a proper way to identify those new perf events. Let's discuss here:)

**Comment by jiere** [comment]
There is a misunderstanding, @marceloamaral. The difficulty we are facing is not retrieving the specific event's counter, but the counter delta and its related application pid relationship establishment. Since the AMX events' execution is asynchronous to the application process, the current delta calculation methodology in task_switch hook is not applicable to the case.

**Comment by marceloamaral** [comment]
Got it @jiere...

Does perf counts the AMX events per processes? If yes, can you check the implementation to see how they do it?
Is there any kernel functions that is called when a processes start/stop using the AMX?
